examples/structured_data/collaborative_filtering_movielens.py

```diff
@@ -0,0 +1,217 @@
+
+df = pd.read_csv(ratings_file)
+
+"""
+First, need to perform some preprocessing to encode users and movies to specific indices.
+This is done as the movieids and userids are non-sequential in nature.
+"""
+
+user_enc = LabelEncoder()
+df["user"] = user_enc.fit_transform(df["userId"].values)
+n_users = df["user"].nunique()
+movie_enc = LabelEncoder()
+df["movie"] = movie_enc.fit_transform(df["movieId"].values)
+n_movies = df["movie"].nunique()
+df["rating"] = df["rating"].values.astype(np.float32)
+min_rating = min(df["rating"])
+max_rating = max(df["rating"])
+print(
+    "Number of users: {}, Number of Movies: {}, Min rating: {}, Max rating: {}".format(
+        n_users, n_movies, min_rating, max_rating
+    )
+)
+
+"""
+## Prepare training and validation data
+"""
+
+X = df[["user", "movie"]].values
+y = df["rating"].values
+X_train, X_test, y_train, y_test = train_test_split(
+    X, y, test_size=0.1, random_state=42
+)
+print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)
+
+"""
+## Create the model
+
+Let's fix some parameters in the architecture. We will use user embeddings and movie
+embeddings, let's keep the dimensions as 50 for both.
+"""
+EMBEDDING_SIZE = 50
+
+
+class EmbeddingLayer(layers.Layer):
+    def __init__(self, n_items, n_factors):
+        super(EmbeddingLayer, self).__init__()
+        self.n_items = n_items
+        self.n_factors = n_factors
+        self.emb = Embedding(
+            self.n_items,
+            self.n_factors,
+            embeddings_initializer="he_normal",
+            embeddings_regularizer=l2(1e-6),
+        )
+
+    def call(self, x):
+        x = self.emb(x)
+        x = Reshape((self.n_factors,))(x)
+        return x
+
+
+class RecommenderNet(keras.Model):
+    def __init__(self, n_users, n_movies, embedding_size, min_rating, max_rating):
+        super(RecommenderNet, self).__init__(name="recnet")
+        self.n_users = n_users
+        self.n_movies = n_movies
+        self.embedding_size = embedding_size
+        self.min_rating = min_rating
+        self.max_rating = max_rating
+        self.user_embedding = Embedding(
+            n_users,
+            embedding_size,
+            embeddings_initializer="he_normal",
+            embeddings_regularizer=l2(1e-6),
+        )
+        self.user_bias = Embedding(n_users, 1)
+        self.movie_embedding = Embedding(
+            n_movies,
+            embedding_size,
+            embeddings_initializer="he_normal",
+            embeddings_regularizer=l2(1e-6),
+        )
+        self.movie_bias = Embedding(n_movies, 1)
+
+    def call(self, inputs):
+        u = self.user_embedding(inputs[:, 0])
+        ub = self.user_bias(inputs[:, 0])
+        m = self.movie_embedding(inputs[:, 1])
+        mb = self.movie_bias(inputs[:, 1])
+        x = Dot(axes=1)([u, m])
+        x = Add()([x, ub, mb])
+        x = Activation("sigmoid")(x)
+        x = Lambda(lambda x: x * (self.max_rating - self.min_rating) + self.min_rating)(
```

Comment:
You don't need to use a Lambda layer, you can just in-line the code:
```
x = x * (self.max_rating - self.min_rating) + self.min_rating
```

Reply:
Can you explain the purpose of this line?

Reply:
The scores are between 0.5 and 5 (minimum and maximum ratings, respectively). I use a sigmoid layer to get a number between 0 and 1, and then scale it back using the minimum and maximum ratings.

Reply:
Better to scale the target ratings than the network's outputs (it's a lot less work and it makes the loss range more typical)